scico/functional/_tvnorm.py

```diff
@@ -0,0 +1,138 @@
+                which to apply the finite difference operator. If
+                ``None``, differences are evaluated along all axes.
+        """
+        self.ndims = ndims
+        self.h0 = snp.array([1.0, 1.0]) / snp.sqrt(2.0)   # lowpass filter
+        self.h1 = snp.array([1.0, -1.0]) / snp.sqrt(2.0)  # highpass filter
+        self.l1norm = L1Norm()
+        self.G = None
+        self.W = None
+
+    def __call__(self, x: Array) -> float:
+        r"""Compute the anisotropic TV norm of an array."""
+        if self.G is None or self.G.shape[1] != x.shape:
+            if self.ndims is None:
+                ndims = x.ndim
+            else:
+                ndims = self.ndims
+            axes = tuple(range(ndims))
+            self.G = FiniteDifference(x.shape, input_dtype=x.dtype, axes=axes, circular=True)
+        return snp.sum(snp.abs(self.G @ x))
+
+    @staticmethod
+    def _shape(idx: int, ndims: int) -> Tuple:
+        """Construct a shape tuple.
+
+        Construct a tuple of size `ndims` with all unit entries except
+        for index `idx`, which has a -1 entry.
+        """
+        return (1,) * idx + (-1,) + (1,) * (ndims - idx - 1)
+
+    def prox(self, v: Array, lam: float = 1.0, **kwargs) -> Array:
+        r"""Approximate proximal operator of the anisotripic  TV norm.
+
+        Approximation of the proximal operator of the anisotropic TV norm,
+        computed via the method described in :cite:`kamilov-2016-parallel`.
+
+        Args:
+            v: Input array :math:`\mb{v}`.
+            lam: Proximal parameter :math:`\lam`.
+            kwargs: Additional arguments that may be used by derived
+                classes.
+        """
+        if self.ndims is None:
+            ndims = v.ndim
+        else:
+            ndims = self.ndims
+        K = 2 * ndims
+
+        if self.W is None or self.W.shape[1] != v.shape:
+            C0 = VerticalStack(  # Stack of lowpass filter operators for each axis
+                [
+                    CircularConvolve(
+                        self.h0.reshape(AnisotropicTVNorm._shape(k, ndims)),
+                        v.shape,
+                        ndims=self.ndims,
+                    )
+                    for k in range(ndims)
+                ]
+            )
+            C1 = VerticalStack(  # Stack of highpass filter operators for each axis
+                [
+                    CircularConvolve(
+                        self.h1.reshape(AnisotropicTVNorm._shape(k, ndims)),
+                        v.shape,
+                        ndims=self.ndims,
+                    )
+                    for k in range(ndims)
+                ]
+            )
+            # single-level shift-invariant Haar transform
+            self.W = VerticalStack((C0, C1), jit=True)
+
+        Wv = self.W @ v
+        # Apply 𝑙1 shrinkage to highpass component of shift-invariant Haar transform
+        Wv = Wv.at[1].set(self.l1norm.prox(Wv[1], snp.sqrt(2) * K * lam))
```

Comment:
I don't see how the proximal of the L1 Norm equivalent to the shrinkage operator we had defined in earlier implementation, which was taken the maximum over the difference from zero and the threshold?

Reply:
According to my reading of the paper, the recommended shrinkage function is the prox of the $\ell_1$ norm. (A more standard form would be to replace $y / |y|$ with $\text{sign}(y)$.)

Reply:
Yes that's correct. My question was more about the implementation of prox of the $\ell_1$ norm in `scico`. How is that equivalent to  $`\max \left( \left| {y} \right| - \lambda, 0 \right)\  \text{sign} \left( {y} \right) `$? In particular, how is [line 101](https://github.com/shnaqvi/scico/blob/tv-norm-alt-ver/scico/functional/_norm.py#L101) (`tmp = 0.5 * (tmp + snp.abs(tmp))`, equivalent to `tmp = snp.max(tmp, 0)` operation with 0?  That's what I'm not seeing. Thanks!

Reply:
Yes, that is rather opaque. It works because $(z + |z|)/2 = \text{max}(z, 0) = (z)_+$. If $z \geq 0$ then $|z| = z$ so $(z + |z|)/2 = z$, and if $z < 0$ then $|z| = -z$ so $(z + |z|)/2 = 0$.